[libbeat] Cache processor docs and memory fixes.

marc-gr opened the pull request:
<!-- Type of change
Please label this PR with one of the following labels, depending on the scope of your change:
- Bug
- Enhancement
- Breaking change
- Deprecation
- Cleanup
- Docs
-->

## Proposed commit message

<!-- Mandatory
Explain here the changes you made on the PR.

Please explain:

- WHAT: patterns used, algorithms implemented, design architecture, message processing, etc.
- WHY:  the rationale/motivation for the changes

This text will be pasted into the squash dialog when the change is committed and will be
a long term historical record of the change to help future contributors understand the
change, please help them by making it clear and comprehensive, they may be you.

If the commit title is adequate to describe both of these things, The text here may be omitted
or replaced with "See title". The title of the PR will be used as the commit message title when
the merge is made and the "See title" marker will be removed if present.

The text here and the PR title will be subject to the PR review process.
-->

- Change cache processor documentation from `write_period` to `write_interval`.
- Fix expiries heap cleanup on partial file writes: popping the expiries heap completely on every partial write caused a panic on the following put operations at the time of expiry check.
- Fix expiries infinite growth when large TTLs and recurring keys are cached: when TTLs are large and keys are recurrent, we kept track of an arbitrary amount of expiry heap entries for the same key, causing memory to grow.

## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/heartbeat/builds/2891)
* Commit: 47866f56ffa79c1850174c6263e36309629f3ca5

### Failed CI Steps
* [:pipeline::arrow_up: Upload Pipeline: .buildkite/heartbeat/heartbeat-pipeline.yml](https://buildkite.com/elastic/heartbeat/builds/2891#018e66c5-d579-4871-8e78-1ebbc7c6710b)

cc @marc-gr

<!--buildkite-pr-comment-heartbeat
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/heartbeat/builds/2891","number":2891,"commit":"47866f56ffa79c1850174c6263e36309629f3ca5"}],"number":2891}
buildkite-pr-comment-->

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/filebeat/builds/4248)
* Commit: 47866f56ffa79c1850174c6263e36309629f3ca5

### Failed CI Steps
* [:pipeline::arrow_up: Upload Pipeline: .buildkite/filebeat/filebeat-pipeline.yml](https://buildkite.com/elastic/filebeat/builds/4248#018e66c5-b8ae-429d-8616-2d927a7b2072)

cc @marc-gr

<!--buildkite-pr-comment-filebeat
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/filebeat/builds/4248","number":4248,"commit":"47866f56ffa79c1850174c6263e36309629f3ca5"}],"number":4248}
buildkite-pr-comment-->

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/beats-metricbeat/builds/3813)
* Commit: 47866f56ffa79c1850174c6263e36309629f3ca5

### Failed CI Steps
* [:pipeline::arrow_up: Upload Pipeline: .buildkite/metricbeat/pipeline.yml](https://buildkite.com/elastic/beats-metricbeat/builds/3813#018e66c5-c67c-4171-8001-417cab234243)

cc @marc-gr

<!--buildkite-pr-comment-beats-metricbeat
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/beats-metricbeat/builds/3813","number":3813,"commit":"47866f56ffa79c1850174c6263e36309629f3ca5"}],"number":3813}
buildkite-pr-comment-->

elasticmachine commented:
## :broken_heart: Build Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-38561/detail/PR-38561/2//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-38561/detail/PR-38561/2//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-38561/detail/PR-38561/2//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-38561/detail/PR-38561/2//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://beats_bk_38561.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/beats-ci/transactions/view?rangeFrom=2024-03-22T15:17:41.478Z&rangeTo=2024-03-22T15:37:41.478Z&transactionName=BUILD+Beats%2Fbeats%2FPR-%7Bnumber%7D&transactionType=job&latencyAggregationType=avg&traceId=1cd10592402185859be292543ebc2363&transactionId=19a5df68425735ee)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats





* Duration: 147 min 53 sec





</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->














### Pipeline error [![1](https://img.shields.io/badge/1%20-red)](https://beats-ci.elastic.co/job/Beats/job/beats/job/PR-38561/2//console)

> This error is likely related to the pipeline itself. Click <a href="https://beats-ci.elastic.co/job/Beats/job/beats/job/PR-38561/2/console">here</a>
> and then you will see the error (either incorrect syntax or an invalid configuration).



## :grey_exclamation: Flaky test report
No test was executed to be analysed.



## :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.


- `/package` : Generate the packages and run the E2E tests.


- `/beats-tester` : Run the installation tests with beats-tester.


- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)




</p>
</details>

<!--COMMENT_GENERATED_WITH_ID_comment.id-->

marc-gr commented:
/test

efd6 reviewed (commented):
I think interval is better term here. Also see [this comment](https://github.com/elastic/beats/pull/36686#discussion_r1339391422) from the addition. Suggest changing the documentation rather than the code.

efd6 reviewed (dismissed):
(hold for discusssion)

marc-gr commented:
> I think interval is better term here. Also see [this comment](https://github.com/elastic/beats/pull/36686#discussion_r1339391422) from the addition. Suggest changing the documentation rather than the code.

That was my initial option, too. The thing that made it for period at the end was mainly that anyone using this setting will have it set up as `write_period`, and by fixing it in that direction would fix functionality for free instead of requiring action of fixing their configurations. If we are okay with this, I am fine fixing the docs instead.

marc-gr commented:
Found two more issues with the processor, one causing panics and the other an unnecessary memory allocation.

marc-gr commented:
/test

efd6 commented on libbeat/processors/cache/file_store.go:
```diff
@@ -287,8 +287,8 @@ func (c *fileStore) writeState(final bool) {
 	now := time.Now()
-	for c.expiries.Len() != 0 {
-		e := c.expiries.pop()
+	for i := 0; i < c.expiries.Len(); i++ {
+		e := c.expiries[i]
```
```suggestion
	for _, e := range c.expiries {
```
The previously logic was obviously written when we were only planning to write out on exit, so the iterator was destructive. It would be nice if the write out also dropped expired elements, but that cannot happen non-destructively within this iterator. It could be done by a mark during this iteration, followed by a sweep in the postamble.

Not for now, but this would look like
```diff
diff --git a/libbeat/processors/cache/file_store.go b/libbeat/processors/cache/file_store.go
index d3820600ac..062a64a916 100644
--- a/libbeat/processors/cache/file_store.go
+++ b/libbeat/processors/cache/file_store.go
@@ -287,10 +287,12 @@ func (c *fileStore) writeState(final bool) {
        enc := json.NewEncoder(f)
        enc.SetEscapeHTML(false)
        now := time.Now()
-       for c.expiries.Len() != 0 {
-               e := c.expiries.pop()
+       var expired int
+       for _, e := range c.expiries {
                if e.Expires.Before(now) {
-                       // Don't write expired elements.
+                       // Don't write expired elements; mark and continue.
+                       e.Expires = time.Time{}
+                       expired++
                        continue
                }
                err = enc.Encode(e)
@@ -299,6 +301,17 @@ func (c *fileStore) writeState(final bool) {
                        return
                }
        }
+       if expired != 0 {
+               // Fix up heap and sweep cache.
+               heap.Init(&c.expiries)
+               for n := 0; (c.effort <= 0 || n < c.effort) && len(c.cache) != 0; n++ {
+                       if c.expiries[0].Expires.After(now) {
+                               break
+                       }
+                       e := c.expiries.pop()
+                       delete(c.cache, e.Key)
+               }
+       }
        // Only mark as not dirty if we succeeded in the write.
        c.dirty = false
        c.log.Infow("write state to file succeeded", "id", c.id, "path", c.path)
```

marc-gr replied:
Not sure this is super worth it, does not this operation of dropping also happen at every Put?

marc-gr replied:
PS: not really related, but I noticed the file write can lead to some issues when the cache is really large, since it stops the world until done due to the need of locking the cache. Not sure about how to avoid it, but just pointing it out as it could cause issues if write_interval is on the low end and cache size is on the larger.

efd6 replied:
> Not sure this is super worth it, does not this operation of dropping also happen at every Put?

It does, but only "just enough". This takes a small amount of time relative to the file write to minimize the memory usage.

> PS: not really related, but I noticed the file write can lead to some issues when the cache is really large, since it stops the world until done due to the need of locking the cache. Not sure about how to avoid it, but just pointing it out as it could cause issues if write_interval is on the low end and cache size is on the larger.

If this does become an issue, the approach would be to:
```
try lock write
return if unsuccess
defer unlock write

go:
    lock cache
    obtain worklist of entries more recent than the expiry time
    for each worklist item:
        write element if still more recent than noted time of heap front
    update note of time of current heap front
    unlock cache
    for each remaining worklist item: # or maybe a small set of worklist
        lock cache
        write element if still more recent than expiry time
        unlock cache
```
This requires adding a new field to the `memStore` type that keeps a track of the front of the heap for this purpose, and a lock to `fileStore`.

The approach here will potentially write out deleted items, but should never write out invalid items.

However, it's significantly more complex and unless there is actually an issue, I don't think this is worth it; tuning should be a first option.

marc-gr replied:
> It does, but only "just enough". This takes a small amount of time relative to the file write to minimize the memory usage.

The only thing a bit odd to me, unless I got something wrong,  is that this memory benefit only applies if using the file write intervals, while when using just memory backend or file without the write intervals, this would just be skipped and the "just enough" effort would be done. That is why I mention that I am not sure is worth it, unless we want to do it somewhere else to benefit all configurations.

efd6 replied:
Yeah, I'm fine with things how they are; the code looks good to me. The only reason I suggested it is that the most common (and likely recommended use would be file-backed) and the cost of this work is small in comparison to the cost of the file write, so it was a reasonable place to insert the work in a way that is routinely run.

efd6 commented:
Failures are unrelated.

Adding this test case to would be good.
```diff
diff --git a/libbeat/processors/cache/mem_store_test.go b/libbeat/processors/cache/mem_store_test.go
index 4a6cf500e0..8d372ab4d7 100644
--- a/libbeat/processors/cache/mem_store_test.go
+++ b/libbeat/processors/cache/mem_store_test.go
@@ -335,6 +335,103 @@ var memStoreTests = []struct {
                        },
                },
        },
+       {
+               name: "re-hit",
+               cfg: config{
+                       Store: &storeConfig{
+                               Memory:   &memConfig{"test"},
+                               Capacity: 1000,
+                               Effort:   10,
+                       },
+                       Get: &getConfig{},
+               },
+               want: &memStore{
+                       id:    "test",
+                       cache: map[string]*CacheEntry{},
+                       refs:  1,
+                       // TTL, capacity and effort are set only by put.
+                       ttl:    -1,
+                       cap:    -1,
+                       effort: -1,
+               },
+               steps: []memStoreTestSteps{
+                       0: {
+                               doTo: func(s *memStore) error {
+                                       putCfg := config{
+                                               Store: &storeConfig{
+                                                       Memory:   &memConfig{"test"},
+                                                       Capacity: 1000,
+                                                       Effort:   10,
+                                               },
+                                               Put: &putConfig{
+                                                       TTL: ptrTo(time.Second),
+                                               },
+                                       }
+                                       s.add(putCfg)
+                                       return nil
+                               },
+                               want: &memStore{
+                                       id:     "test",
+                                       cache:  map[string]*CacheEntry{},
+                                       refs:   2,
+                                       dirty:  false,
+                                       ttl:    time.Second,
+                                       cap:    1000,
+                                       effort: 10,
+                               },
+                       },
+                       1: {
+                               doTo: func(s *memStore) error {
+                                       s.Put("one", 1)
+                                       s.Put("two", 2)
+                                       s.Put("three", 3)
+                                       return nil
+                               },
+                               want: &memStore{
+                                       id: "test",
+                                       cache: map[string]*CacheEntry{
+                                               "one":   {Key: "one", Value: int(1), index: 0},
+                                               "two":   {Key: "two", Value: int(2), index: 1},
+                                               "three": {Key: "three", Value: int(3), index: 2},
+                                       },
+                                       expiries: expiryHeap{
+                                               {Key: "one", Value: int(1), index: 0},
+                                               {Key: "two", Value: int(2), index: 1},
+                                               {Key: "three", Value: int(3), index: 2},
+                                       },
+                                       refs:   2,
+                                       dirty:  true,
+                                       ttl:    time.Second,
+                                       cap:    1000,
+                                       effort: 10,
+                               },
+                       },
+                       2: {
+                               doTo: func(s *memStore) error {
+                                       s.Put("two", 2)
+                                       return nil
+                               },
+                               want: &memStore{
+                                       id: "test",
+                                       cache: map[string]*CacheEntry{
+                                               "one":   {Key: "one", Value: int(1), index: 1},
+                                               "two":   {Key: "two", Value: int(2), index: 0},
+                                               "three": {Key: "three", Value: int(3), index: 2},
+                                       },
+                                       expiries: expiryHeap{
+                                               {Key: "two", Value: int(2), index: 0},
+                                               {Key: "one", Value: int(1), index: 1},
+                                               {Key: "three", Value: int(3), index: 2},
+                                       },
+                                       refs:   2,
+                                       dirty:  true,
+                                       ttl:    time.Second,
+                                       cap:    1000,
+                                       effort: 10,
+                               },
+                       },
+               },
+       },
 }
 
 func TestMemStore(t *testing.T) {
```
This fails without the changes to `memStore` with
```
--- FAIL: TestMemStore (0.00s)
    --- FAIL: TestMemStore/re-hit (0.00s)
        mem_store_test.go:460: unexpected memStore step 2 result:
            --- want
            +++ got
              &cache.memStore{
              	... // 1 ignored field
              	cache: map[string]*cache.CacheEntry{
              		"one": &{
              			... // 1 ignored and 2 identical fields
            - 			index: 1,
            + 			index: 0,
              		},
              		"three": &{Key: "three", Value: int(3), index: 2},
              		"two": &{
              			... // 1 ignored and 2 identical fields
            - 			index: 0,
            + 			index: 3,
              		},
              	},
              	expiries: cache.expiryHeap{
              		&{
            - 			Key:   "two",
            + 			Key:   "one",
            - 			Value: int(2),
            + 			Value: int(1),
              			... // 1 ignored and 1 identical fields
              		},
              		&{
            - 			Key:   "one",
            + 			Key:   "two",
            - 			Value: int(1),
            + 			Value: int(2),
              			... // 1 ignored and 1 identical fields
              		},
              		&{Key: "three", Value: int(3), index: 2},
            + 		&{
            + 			Key:     "two",
            + 			Value:   int(2),
            + 			Expires: s"2024-03-26 07:03:04.463805506 +1030 ACDT m=+1.013529260",
            + 			index:   3,
            + 		},
              	},
              	ttl:  s"1s",
              	refs: 2,
              	... // 4 identical fields
              }
FAIL
```
and this test for `fileStore`
```diff
diff --git a/libbeat/processors/cache/file_store_test.go b/libbeat/processors/cache/file_store_test.go
index 22d43083e0..163899eca0 100644
--- a/libbeat/processors/cache/file_store_test.go
+++ b/libbeat/processors/cache/file_store_test.go
@@ -352,6 +352,109 @@ var fileStoreTests = []struct {
                        {Key: "three", Value: 3.0},
                },
        },
+       {
+               name: "periodic_write",
+               cfg: config{
+                       Store: &storeConfig{
+                               File:     &fileConfig{ID: "test"},
+                               Capacity: 1000,
+                               Effort:   10,
+                       },
+                       Get: &getConfig{},
+               },
+               want: &fileStore{path: "testdata/periodic_write", memStore: memStore{
+                       id:    "test",
+                       cache: map[string]*CacheEntry{},
+                       refs:  1,
+                       // TTL, capacity and effort are set only by put.
+                       ttl:    -1,
+                       cap:    -1,
+                       effort: -1,
+               }},
+               steps: []fileStoreTestSteps{
+                       0: {
+                               doTo: func(s *fileStore) error {
+                                       putCfg := config{
+                                               Store: &storeConfig{
+                                                       File:     &fileConfig{ID: "test"},
+                                                       Capacity: 1000,
+                                                       Effort:   10,
+                                               },
+                                               Put: &putConfig{
+                                                       TTL: ptrTo(time.Second),
+                                               },
+                                       }
+                                       s.add(putCfg)
+                                       return nil
+                               },
+                               want: &fileStore{path: "testdata/periodic_write", memStore: memStore{
+                                       id:     "test",
+                                       cache:  map[string]*CacheEntry{},
+                                       refs:   2,
+                                       dirty:  false,
+                                       ttl:    time.Second,
+                                       cap:    1000,
+                                       effort: 10,
+                               }},
+                       },
+                       1: {
+                               doTo: func(s *fileStore) error {
+                                       s.Put("one", 1)
+                                       s.Put("two", 2)
+                                       s.Put("three", 3)
+                                       return nil
+                               },
+                               want: &fileStore{path: "testdata/periodic_write", memStore: memStore{
+                                       id: "test",
+                                       cache: map[string]*CacheEntry{
+                                               "one":   {Key: "one", Value: int(1), index: 0},
+                                               "two":   {Key: "two", Value: int(2), index: 1},
+                                               "three": {Key: "three", Value: int(3), index: 2},
+                                       },
+                                       expiries: expiryHeap{
+                                               {Key: "one", Value: int(1), index: 0},
+                                               {Key: "two", Value: int(2), index: 1},
+                                               {Key: "three", Value: int(3), index: 2},
+                                       },
+                                       refs:   2,
+                                       dirty:  true,
+                                       ttl:    time.Second,
+                                       cap:    1000,
+                                       effort: 10,
+                               }},
+                       },
+                       2: {
+                               doTo: func(s *fileStore) error {
+                                       s.writeState(false)
+                                       return nil
+                               },
+                               want: &fileStore{path: "testdata/periodic_write", memStore: memStore{
+                                       id: "test",
+                                       cache: map[string]*CacheEntry{
+                                               "one":   {Key: "one", Value: int(1), index: 0},
+                                               "two":   {Key: "two", Value: int(2), index: 1},
+                                               "three": {Key: "three", Value: int(3), index: 2},
+                                       },
+                                       expiries: expiryHeap{
+                                               {Key: "one", Value: int(1), index: 0},
+                                               {Key: "two", Value: int(2), index: 1},
+                                               {Key: "three", Value: int(3), index: 2},
+                                       },
+                                       refs:   2,
+                                       dirty:  false,
+                                       ttl:    time.Second,
+                                       cap:    1000,
+                                       effort: 10,
+                               }},
+                       },
+               },
+               wantPersisted: []*CacheEntry{
+                       // Numeric values are float due to JSON round-trip.
+                       {Key: "one", Value: 1.0},
+                       {Key: "two", Value: 2.0},
+                       {Key: "three", Value: 3.0},
+               },
+       },
 }
 
 func TestFileStore(t *testing.T) {
```
otherwise failing with
```
--- FAIL: TestFileStore (0.00s)
    --- FAIL: TestFileStore/periodic_write (0.00s)
        file_store_test.go:497: unexpected fileStore step 2 result:
            --- want
            +++ got
              &cache.fileStore{
              	memStore: cache.memStore{
              		... // 1 ignored field
              		cache: map[string]*cache.CacheEntry{
              			"one": &{
              				... // 1 ignored and 2 identical fields
            - 				index: 0,
            + 				index: -1,
              			},
              			"three": &{
              				... // 1 ignored and 2 identical fields
            - 				index: 2,
            + 				index: -1,
              			},
              			"two": &{
              				... // 1 ignored and 2 identical fields
            - 				index: 1,
            + 				index: -1,
              			},
              		},
              		expiries: cache.expiryHeap{
            - 			&{Key: "one", Value: int(1)},
            - 			&{Key: "two", Value: int(2), index: 1},
            - 			&{Key: "three", Value: int(3), index: 2},
              		},
              		ttl:  s"1s",
              		refs: 2,
              		... // 4 identical fields
              	},
              	path: "testdata/periodic_write",
              	... // 2 ignored fields
              }
FAIL
```

marc-gr commented on libbeat/processors/cache/file_store_test.go:
```diff
@@ -410,7 +410,7 @@ func TestFileStore(t *testing.T) {
 				err = dec.Decode(&e)
 				if err != nil {
-					if err != io.EOF {
+					if !errors.Is(err, io.EOF) {
```
linter was complaining, either we add a nolint or use this I guess

efd6 replied:
Or just ignore the linter

marc-gr commented:
Removed linter changes, not sure if we just want to ignore them or adding the nolint is preferred over fixing them.

Used heap.Fix and added related tests.

Note about the tests suggested, changed last step to see a change in the heap root, since heap Fix will not "properly" sort the slice, as it will only take care of the heap invariant `!h.Less(j, i) for 0 <= i < h.Len() and 2*i+1 <= j <= 2*i+2 and j < h.Len()` and the suggested change did not reflect in any changes in the underlying slice order.

efd6 commented:
> Removed linter changes, not sure if we just want to ignore them or adding the nolint is preferred over fixing them.

I would prefer to just ignore them, but Andrew may disagree.

> Note about the tests suggested, changed last step to see a change in the heap root, since heap Fix will not "properly" sort the slice, as it will only take care of the heap invariant !h.Less(j, i) for 0 <= i < h.Len() and 2*i+1 <= j <= 2*i+2 and j < h.Len() and the suggested change did not reflect in any changes in the underlying slice order.

Heap sort is not lexical sort, so this is fine. We don't care about the order of the heap entries, just that the heap behaves like a heap.

efd6 commented:
/test

marc-gr commented:
/test

efd6 commented:
/test

andrewkroh commented:
I think the failures in metricbeat-pythonIntegTest are caused by https://github.com/elastic/beats/issues/38539.

marc-gr commented:
@Mergifyio backport 8.12
